### PR TITLE
[Merged by Bors] - p2p address book refactor to separate package

### LINKED ISSUE #3132: p2p: stale records in the address book
There are several problems with address book that we are using:

1. Peers that are long gone are not removed from the address book.
2. If local address changes (for example due to nat update) we do not notify our peers in the network about it.

Both issues can be handled by periodically exchanging discovery data with N peers from the address book. If peers doesn't respond for T attempts it must be removed from the local address book. Peer is expected to update local addresses based on the received data, but this needs to be tested.

Additionally, address book code was copied from somewhere, and then it got borked. Many original features are not used and the code itself doesn't look good. Would be nice to rewrite it properly before mainnet.

### PR DESCRIPTION
**Pull request by abergasov**
## Motivation
Closes #3132 

## Changes
move addressbook to separate package with some refactoring

## Test Plan
unit tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: log/zap.go [hunk 1]
```diff
@@ -36,6 +36,11 @@ func (l Log) Error(format string, args ...interface{}) {
 	l.logger.Sugar().Errorf(format, args...)
 }
 
+// LogError prints message with fields.
+func (l Log) LogError(msg string, err error, fields ...LoggableField) {
```
Comment (dshulyak) [inline log/zap.go]: please just wrap the error as log.Err field and remove this method

File: log/zap.go [hunk 2]
```diff
@@ -278,6 +283,11 @@ func (fl FieldLogger) Error(msg string, fields ...LoggableField) {
 	fl.l.Error(msg, unpack(append(fields, String("name", fl.name)))...)
 }
 
+// LogError prints message with fields.
+func (fl FieldLogger) LogError(msg string, err error, fields ...LoggableField) {
+	fl.l.Error(msg, unpack(append(fields, Err(err), String("name", fl.name)))...)
```
Comment (dshulyak) [inline log/zap.go]: same

File: p2p/addressbook/addrbook.go [hunk 3]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+	}
+	if len(nonEmptyBuckets) == 0 {
+		return nil // no addresses to extract.
+	}
+	large := 1 << 30
+	factor := 1.0
+	for {
+		// pick a random bucket.
+		bucket := nonEmptyBuckets[a.rand.Intn(len(nonEmptyBuckets))]
+		if len(buckets[bucket]) == 0 {
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go]: how it can be zero if it was filtered out above?

File: p2p/addressbook/addrbook.go [hunk 4]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+
+		// Pick a random entry in it.
+		var ka *knownAddress
+		nth := a.rand.Intn(len(buckets[bucket]))
+		for _, value := range buckets[bucket] {
+			if nth == 0 {
+				ka = value
+			}
+			nth--
+		}
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go:267]: why ka is not buckets[bucket][nth] ? maybe i just dont understand this loop

File: p2p/addressbook/addrbook.go [hunk 5]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+		if ka == nil {
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go]: why it can be nil?

File: p2p/addressbook/addrbook.go [hunk 6]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+	if numAddresses > a.cfg.GetAddrMax {
+		numAddresses = a.cfg.GetAddrMax
 	} else if numAddresses == 0 {
 		numAddresses = len(allAddr)
 	}
 
-	// Fisher-Yates shuffle the array. We only need to do the first
-	// `numAddresses' since we are throwing the rest.
+	result := make([]*AddrInfo, 0, numAddresses)
+	a.mu.RLock()
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go:290]: it is preferable to always defer unlock for safety if logic allows

File: p2p/addressbook/addrbook.go [hunk 7]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+
+	if len(a.lastAnchorPeers) == 0 {
+		return addresses
+	}
+
+	anchorPeers := make([]*knownAddress, len(a.lastAnchorPeers), len(a.lastAnchorPeers))
+	copy(anchorPeers, a.lastAnchorPeers)
+	for i := 0; i < a.cfg.AnchorPeersCount; i++ {
+		j := rand.Intn(len(anchorPeers)-i) + i // pick a number between current index and the end.
+		anchorPeers[i], anchorPeers[j] = anchorPeers[j], anchorPeers[i]
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go:322]: are you shuffling? there is a function rand.Shuffle

File: p2p/addressbook/addrbook.go [hunk 8]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+func (a *AddrBook) BootstrapAddressCache() []*AddrInfo {
+	addresses := a.AddressCache()
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	if len(a.lastAnchorPeers) == 0 {
+		return addresses
+	}
+
+	anchorPeers := make([]*knownAddress, len(a.lastAnchorPeers), len(a.lastAnchorPeers))
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go]: why do you need to specify both length and capacity if they are equal?

File: p2p/addressbook/addrbook.go [hunk 9]
```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
...
+	copy(anchorPeers, a.lastAnchorPeers)
+	for i := 0; i < a.cfg.AnchorPeersCount; i++ {
+		j := rand.Intn(len(anchorPeers)-i) + i // pick a number between current index and the end.
```
Comment (dshulyak) [inline p2p/addressbook/addrbook.go]: why a.rand is not used here?

**Review by dshulyak — CHANGES_REQUESTED** [review]

File: p2p/addressbook/addrbook.go (hunk 3 above)
Comment (abergasov) [reply p2p/addressbook/addrbook.go]: forget to drop)

File: p2p/addressbook/addrbook.go (hunk 9 above)
Comment (abergasov) [reply p2p/addressbook/addrbook.go]: replaced to shuffle

File: p2p/addressbook/addrbook.go (hunk 4 above)
Comment (abergasov) [reply p2p/addressbook/addrbook.go:267]: structure is `map[peer.ID]*knownAddress`, so we should iterate in loop, cause rand return only int

File: p2p/addressbook/addrbook.go (hunk 5 above)
Comment (abergasov) [reply p2p/addressbook/addrbook.go]: forget to drop 2)

**Review by dshulyak — APPROVED** [review]

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors merge

**Pull request closed without merging**